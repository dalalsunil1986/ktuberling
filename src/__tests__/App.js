import React from 'react';
import renderer from 'react-test-renderer';
import xhr from 'xhr';
import { error as logError } from 'extra-log';

import App from '../App';

jest.mock('xhr', () => jest.fn());
jest.mock('extra-log', () => ({ error: jest.fn() }));

jest.mock('../app.less', () => ({}));
jest.mock('../components/Brand', () => 'Brand');
jest.mock('../components/Busy', () => 'Busy');
jest.mock('../components/NavbarButton', () => 'NavbarButton');
jest.mock('../components/Playground', () => 'Playground');

const findById = (id, json) => {
  let node = null;

  if (json.props.id === id) {
    node = json;
  } else if (json.children) {
    for (let ind = json.children.length - 1; 0 <= ind; --ind) {
      node = findById(id, json.children[ind]);
      if (node) {
        break;
      }
    }
  }

  return node;
};

describe('App', () => {
  const playgroundsData = `[
  {"name":"playground0","file":"file0.file","items":[]},
  {"name":"playground1","file":"file1.file","items":[]}
]`;


  beforeEach(() => {
    xhr.mockReset();
    logError.mockReset();
  });


  it('renders', () => {
    const tree = renderer.create(<App />);

    expect(tree).toMatchSnapshot();
  });


  it('loads play grounds and set first', () => {
    const tree = renderer.create(<App />);

    expect(xhr).toHaveBeenCalledTimes(1);
    expect(xhr.mock.calls).toMatchSnapshot();

    const callback = xhr.mock.calls[0][1];
    callback(undefined, undefined, playgroundsData);

    expect(tree).toMatchSnapshot();
  });


  it('loads play grounds and handles error', () => {
    const tree = renderer.create(<App />);

    expect(xhr).toHaveBeenCalledTimes(1);
    expect(xhr.mock.calls).toMatchSnapshot();

    const callback = xhr.mock.calls[0][1];
    callback(true, undefined, playgroundsData);

    expect(logError).toHaveBeenCalledTimes(1);
    expect(tree).toMatchSnapshot();
  });


  it('show playgrounds selection', () => {
    const tree = renderer.create(<App />);

    expect(xhr).toHaveBeenCalledTimes(1);
    expect(xhr.mock.calls).toMatchSnapshot();

    const callback = xhr.mock.calls[0][1];
    callback(undefined, undefined, playgroundsData);

    const li = findById('playgroundsSelector', tree.toJSON());
    li.props.onClick();
    expect(tree).toMatchSnapshot();
  });


  it('handle on playground item click and change playground', () => {
    const tree = renderer.create(<App />);

    expect(xhr).toHaveBeenCalledTimes(1);
    expect(xhr.mock.calls).toMatchSnapshot();

    const callback = xhr.mock.calls[0][1];
    callback(undefined, undefined, playgroundsData);

    const li = findById('playgroundsSelector', tree.toJSON());
    li.props.onClick();


    const clickbleItemInactive = findById('item_1', tree.toJSON());

    clickbleItemInactive.props.onClick();
    expect(tree).toMatchSnapshot('tree');
  });
});
