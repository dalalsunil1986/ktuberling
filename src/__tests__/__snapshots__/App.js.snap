// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App handle on playground item click and change playground 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "uri": "assets/data/playgrounds.json",
    },
    [Function],
  ],
]
`;

exports[`App handle on playground item click and change playground: tree 1`] = `
<div
  className="app"
>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container-fluid"
    >
      <Brand />
      <div
        className="collapse navbar-collapse"
      >
        <ul
          className="nav navbar-nav"
        >
          <NavbarButton
            disabled={true}
            text="New"
          />
          <NavbarButton
            disabled={true}
            text="Save"
          />
          <NavbarButton
            disabled={true}
            text="Print"
          />
          <NavbarButton
            disabled={true}
            text="Undo"
          />
          <NavbarButton
            disabled={true}
            text="Redo"
          />
          <li
            className="dropdown open"
            id="playgroundsSelector"
            onClick={[Function]}
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              role="button"
            >
              Playground 
              <span
                className="caret"
              />
            </a>
            <ul
              className="dropdown-menu"
            >
              <li
                id="item_0"
                onClick={[Function]}
              >
                <a
                  href="#"
                >
                  playground0
                </a>
              </li>
              <li
                id="item_1"
                onClick={[Function]}
              >
                <a
                  href="#"
                >
                  playground1
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <Playground
      ground={
        Object {
          "file": "file1.file",
          "items": Array [],
          "name": "playground1",
        }
      }
    />
    <Busy
      active={false}
    />
  </div>
</div>
`;

exports[`App loads play grounds and handles error 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "uri": "assets/data/playgrounds.json",
    },
    [Function],
  ],
]
`;

exports[`App loads play grounds and handles error 2`] = `
<div
  className="app"
>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container-fluid"
    >
      <Brand />
      <div
        className="collapse navbar-collapse"
      >
        <ul
          className="nav navbar-nav"
        >
          <NavbarButton
            disabled={true}
            text="New"
          />
          <NavbarButton
            disabled={true}
            text="Save"
          />
          <NavbarButton
            disabled={true}
            text="Print"
          />
          <NavbarButton
            disabled={true}
            text="Undo"
          />
          <NavbarButton
            disabled={true}
            text="Redo"
          />
          <li
            className="dropdown"
            id="playgroundsSelector"
            onClick={[Function]}
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              role="button"
            >
              Playground 
              <span
                className="caret"
              />
            </a>
            <ul
              className="dropdown-menu"
            />
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <Playground
      ground={undefined}
    />
    <Busy
      active={true}
    />
  </div>
</div>
`;

exports[`App loads play grounds and set first 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "uri": "assets/data/playgrounds.json",
    },
    [Function],
  ],
]
`;

exports[`App loads play grounds and set first 2`] = `
<div
  className="app"
>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container-fluid"
    >
      <Brand />
      <div
        className="collapse navbar-collapse"
      >
        <ul
          className="nav navbar-nav"
        >
          <NavbarButton
            disabled={true}
            text="New"
          />
          <NavbarButton
            disabled={true}
            text="Save"
          />
          <NavbarButton
            disabled={true}
            text="Print"
          />
          <NavbarButton
            disabled={true}
            text="Undo"
          />
          <NavbarButton
            disabled={true}
            text="Redo"
          />
          <li
            className="dropdown"
            id="playgroundsSelector"
            onClick={[Function]}
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              role="button"
            >
              Playground 
              <span
                className="caret"
              />
            </a>
            <ul
              className="dropdown-menu"
            >
              <li
                id="item_0"
                onClick={[Function]}
              >
                <a
                  href="#"
                >
                  playground0
                </a>
              </li>
              <li
                id="item_1"
                onClick={[Function]}
              >
                <a
                  href="#"
                >
                  playground1
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <Playground
      ground={
        Object {
          "file": "file0.file",
          "items": Array [],
          "name": "playground0",
        }
      }
    />
    <Busy
      active={false}
    />
  </div>
</div>
`;

exports[`App renders 1`] = `
<div
  className="app"
>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container-fluid"
    >
      <Brand />
      <div
        className="collapse navbar-collapse"
      >
        <ul
          className="nav navbar-nav"
        >
          <NavbarButton
            disabled={true}
            text="New"
          />
          <NavbarButton
            disabled={true}
            text="Save"
          />
          <NavbarButton
            disabled={true}
            text="Print"
          />
          <NavbarButton
            disabled={true}
            text="Undo"
          />
          <NavbarButton
            disabled={true}
            text="Redo"
          />
          <li
            className="dropdown"
            id="playgroundsSelector"
            onClick={[Function]}
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              role="button"
            >
              Playground 
              <span
                className="caret"
              />
            </a>
            <ul
              className="dropdown-menu"
            />
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <Playground
      ground={undefined}
    />
    <Busy
      active={true}
    />
  </div>
</div>
`;

exports[`App show playgrounds selection 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "uri": "assets/data/playgrounds.json",
    },
    [Function],
  ],
]
`;

exports[`App show playgrounds selection 2`] = `
<div
  className="app"
>
  <nav
    className="navbar navbar-default"
  >
    <div
      className="container-fluid"
    >
      <Brand />
      <div
        className="collapse navbar-collapse"
      >
        <ul
          className="nav navbar-nav"
        >
          <NavbarButton
            disabled={true}
            text="New"
          />
          <NavbarButton
            disabled={true}
            text="Save"
          />
          <NavbarButton
            disabled={true}
            text="Print"
          />
          <NavbarButton
            disabled={true}
            text="Undo"
          />
          <NavbarButton
            disabled={true}
            text="Redo"
          />
          <li
            className="dropdown open"
            id="playgroundsSelector"
            onClick={[Function]}
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              className="dropdown-toggle"
              data-toggle="dropdown"
              href="#"
              role="button"
            >
              Playground 
              <span
                className="caret"
              />
            </a>
            <ul
              className="dropdown-menu"
            >
              <li
                id="item_0"
                onClick={[Function]}
              >
                <a
                  href="#"
                >
                  playground0
                </a>
              </li>
              <li
                id="item_1"
                onClick={[Function]}
              >
                <a
                  href="#"
                >
                  playground1
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <Playground
      ground={
        Object {
          "file": "file0.file",
          "items": Array [],
          "name": "playground0",
        }
      }
    />
    <Busy
      active={false}
    />
  </div>
</div>
`;
